packages/core/src/State.ts

```diff
@@ -290,9 +290,12 @@ export class State<
    */
   public matches<TSV extends TTypestate['value']>(
     parentStateValue: TSV
-  ): this is TTypestate extends { value: TSV }
-    ? State<TTypestate['context'], TEvent, TStateSchema, TTypestate>
-    : never {
+  ): this is State<
+    (TTypestate extends { value: TSV } ? TTypestate : never)['context'],
+    TEvent,
+    TStateSchema,
+    TTypestate
+  > & { value: TSV } {
```

Comment:
Can't say I understand this but seems to be crucial for the fix 😅

Reply:
Haha yeah 😅 So here's the explanation for this:

If `TTypestate` is just `{ string: TContext }`, then without that `{ value: TSV }`, this would evaluate to:

```ts
this is State<TContext, TEvent, TStateSchema, TTypestate>
```

So if we have `state.matches(...)`, that would evaluate to:

```ts
if (state.matches('foo')) {
  // state is State<TContext, TEvent, TStateSchema, TTypestate>
} else {
  // state is NOT State<TContext, TEvent, TStateSchema, TTypestate>
  // ... which is "never" (see below)
}
```

So "state is NOT `State<TContext, TEvent, TStateSchema, TTypestate>`" evaluates to `never` because TypeScript is basically doing this:

```ts
this extends State<TContext, TEvent, TStateSchema, TTypestate> ? never : State<TContext, TEvent, TStateSchema, TTypestate>;
```

And because `state` is _always_ `State<TContext, TEvent, TStateSchema, TTypestate>` and `A & !A` always evaluates to false, that expands to:

```ts
State<TContext, TEvent, TStateSchema, TTypestate> extends State<TContext, TEvent, TStateSchema, TTypestate>
  ? never
  : State<TContext, TEvent, TStateSchema, TTypestate>;
```

And it will always be `never`... which was a problem.

So we narrowed it down to `State<TContext, TEvent, TStateSchema, TTypestate> & { value: TSV }` so that `this` can still be `State<TContext, TEvent, TStateSchema, TTypestate>`, just with a different `value`.